respy/exogenous_processes.py

```diff
@@ -0,0 +1,102 @@
+
+from respy.config import COVARIATES_DOT_PRODUCT_DTYPE
+from respy.parallelization import parallelize_across_dense_dimensions
+from respy.shared import create_dense_state_space_columns
+
+
+@parallelize_across_dense_dimensions
+def compute_transition_probabilities(states, optim_paras):
+    """Compute the transition probabilities between states due to exogenous processes.
+
+    The exogenous process is defined by a multinomial logit. The steps are
+
+    1. For each realization of each exogenous process, calculate the dot product of
+       process parameters and covariates and compute the transition probabilities with
+       the softmax function.
+
+    2. For each dense index this `_SingleDimStateSpace` may transition to, collect the
+       transition probabilities for each exogenous process and multiply them.
+
+    This function returns a dictionary for each `respy.state_space._SingleDimStateSpace`
+    whose keys are the dense state space indices to where the states can transition to.
+    To each key belongs a probability vector which is the product of probabilities
+    defined by the exogenous processes.
+
+    As an example, assume a model without observables and types but one exogenous
+    processes named "exog_proc" with two possible values, zero and one. The return
+    taking into account the decorator will look like this:
+
+    .. code-block:: python
+
+        {  # Outer dictionary created by deco.
+            (0,): {(0,): prob_vector, (1,): prob_vector},  # Inner dictionary create by
+            (1,): {(0,): prob_vector, (1,): prob_vector},  # the function.
+        }
+
+    """
+    exogenous_processes = optim_paras["exogenous_processes"]
+    dense_columns = create_dense_state_space_columns(optim_paras)
+
+    # Get the dense index of the current states instead of having an additional funcarg.
+    dense_idx = list(states.groupby(dense_columns).groups)[0]
+    dense_idx = (dense_idx,) if np.isscalar(dense_idx) else dense_idx
+
+    # Compute the probabilities for every exogenous process.
+    probabilities = []
+    for exog_proc in exogenous_processes:
+
+        # Create the dot product of covariates and parameters.
+        x_betas = []
+        for params in exogenous_processes[exog_proc].values():
+            x_beta = np.dot(
+                states[params.index].to_numpy(dtype=COVARIATES_DOT_PRODUCT_DTYPE),
+                params.to_numpy(),
+            )
+            x_betas.append(x_beta)
+
+        probs = special.softmax(np.column_stack(x_betas), axis=1)
+        probabilities.append(probs)
+
+    # Assign one probability vector per process to one `_SingleDimStateSpace`. Instead
+    # of storing multiple vectors per state space, multiply them.
+    transition_probabilities = {}
+    comb_exog_procs = itertools.product(
+        *[range(len(levels)) for levels in exogenous_processes.values()]
+    )
+    n_exog_procs = len(exogenous_processes)
+    pos = dense_columns.index(list(exogenous_processes)[0])
```

Comment:
Does this step really make memory usage smaller? It seems like one could potentially create massive arrays here and I don't see the problem of carrying several arrays a priori.

Reply:
Maybe you are right. For one exogenous process, it does not matter. For two independent processes with two potential values each, you would have 4 separate probability vectors and 4 melted probability vectors which does not matter as well. But, for more potential values, say three, you would 6 separate vectors and 9 melted.

Reply:
This was a very good comment. We keep the process-specific probabilities now, and combine them to transition probabilities between dense indices if necessary.